doc/developer/sqllogictest.md

```diff
@@ -90,6 +90,20 @@ CREATE TABLE bar (
 
 ## Materialize-specific behavior in sqllogictest
 
+
+### Server resets
+
+
+By default, sqllogictest will create a single Mz instance and run
+all the SLT tests against it. For tests that require special treatment,
+the `reset-server` can be used to bring about a new, pristine Mz instance.
+
+
+For tests that involve object Ids (such as certain `EXPLAIN` tests and
+those that consult the introspection tables), `reset-server` should
+be used at the start of the .slt file. For tests that manipulate clusters,
```

Comment:
Using reset-server at the end of the test script seems like a recipe for trouble. If we bail out earlier in the script (say, a bug in Materialize), then the _next_ test might have spurious failures, because the server is left in a bad state.

Our options are:
1. Always reset the server after a failure.
2. Look for a reset-server directive after a failure and reset the server once if you see one.
3. Drop all clusters, databases, and users after every test, to start fresh.

Personally I like 3, as it means you don’t have to remember to do something in each test script that manipulates clusters. In general just seems nice to have each SLT test start from as fresh a state as possible, even if we don’t restart the entire server.

Reply:
But #3 is going to bring the execution speed almost back to where it was before, defeating the purpose of the entire optimization. I can do #1 most easily -- and I also think the reset it should happen for all failures regardless of the presence of a reset-server directive elsewhere in the file (so option #2 is actually inferior).

Would that be acceptable?

Reply:
Ohhh right, it’s the slow cluster startup time, not just the envd boot time.

So, yeah, 1 makes sense!

Reply:
Wait, what if we do (3), but we o only drop and recreate the default cluster and its replica if it has the wrong size? I.e., ensuring that there is exactly one default.r1 replica, but not dropping and recreating it unnecessarily.

Reply:
> Wait, what if we do (3), but we o only drop and recreate the default cluster and its replica if it has the wrong size? I.e., ensuring that there is exactly one default.r1 replica, but not dropping and recreating it unnecessarily.

I guess eventually clusters and replicas will have more properties and whatnot than just their name and their SIZE, so it will be difficult to know if we have deviated too much from the pristine state or not, and if our recovery actions have sufficiently returned the cluster back to a known-expected state.

I spent one too many hours in the reactor core that is Rust, so if you would like No 3 implemented this needs to wait for the other day. The solution for No 1 is a one-liner:

```
diff --git a/src/sqllogictest/src/runner.rs b/src/sqllogictest/src/runner.rs
index 1a1740721..29d18eca9 100644
--- a/src/sqllogictest/src/runner.rs
+++ b/src/sqllogictest/src/runner.rs
@@ -673,7 +673,11 @@ impl<'a> Runner<'a> {
                         } else {
                             // If we failed to execute a statement that was supposed to succeed,
                             // running the rest of the tests in this file will probably cause
-                            // false positives, so just give up on the file entirely.
+                            // false positives, so just give up on the file entirely and reset
+                            // the server to prepare for running the next file, if any
+
+                            self.reset().await?;
+
                             Ok(Outcome::Bail {
                                 cause: Box::new(other),
                                 location: location.clone(),
@@ -932,6 +936,7 @@ impl RunnerInner {
                         return Ok(Outcome::Success);
                     }
                 }
+
                 Ok(Outcome::PlanFailure {
                     error: anyhow!(error),
                     location,
```